python/ray/data/_internal/datasource/kafka_datasource.py

```diff
@@ -0,0 +1,568 @@
+                in the remote task without requiring 'self' to be serialized.
+                """
+
+                def kafka_read_fn() -> Iterable[Block]:
+                    """Read function for a single Kafka partition using kafka-python.
+
+                    This function runs remotely in a Ray task. It creates a KafkaConsumer,
+                    reads messages from a single assigned partition, and yields PyArrow tables
+                    incrementally for efficient streaming processing.
+                    """
+                    from kafka import KafkaConsumer, TopicPartition
+
+                    # Build consumer configuration
+                    consumer_config = _build_consumer_config_for_read(
+                        bootstrap_servers, authentication, topic_name, partition_id
+                    )
+
+                    # Create the Kafka consumer
+                    consumer = KafkaConsumer(**consumer_config)
+                    try:
+                        # Assign only the specific partition for this task
+                        topic_partition = TopicPartition(topic_name, partition_id)
+                        consumer.assign([topic_partition])
+
+                        # Determine start offset for this partition
+                        if start_offset is not None:
+                            start_off = _resolve_offset(
+                                consumer, topic_partition, start_offset, is_end=False
+                            )
+                        else:
+                            beginning_offsets = consumer.beginning_offsets(
+                                [topic_partition]
+                            )
+                            start_off = beginning_offsets[topic_partition]
+
+                        # Determine end offset for this partition
+                        if end_offset is not None:
+                            end_off = _resolve_offset(
+                                consumer, topic_partition, end_offset, is_end=True
+                            )
+                        else:
+                            end_offsets = consumer.end_offsets([topic_partition])
+                            end_off = end_offsets[topic_partition]
+
+                        # Validate start_offset <= end_offset
+                        if start_off > end_off:
+                            raise ValueError(
+                                f"start_offset ({start_off}) > end_offset ({end_off}) "
+                                f"for partition {partition_id} in topic {topic_name}"
+                            )
+                        # Seek to the requested starting position
+                        consumer.seek(topic_partition, start_off)
+
+                        records = []
+                        records_read = 0
+                        # Use per_task_row_limit if provided, otherwise use max_records_per_task
+                        ctx = DataContext.get_current()
```

Comment:
Set the datasource in the constructor, and use that. DataContext should be set at composition time (not execution time like the read task) to avoid janky behavior